Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp

```diff
@@ -391,1286 +2515,359 @@ double BytecodeInterpreter::read_value<double>(ReadonlyBytes data)
-        return;
-    case Instructions::f64_const.value():
-        configuration.value_stack().append(Value(instruction.arguments().get<double>()));
-        return;
-    case Instructions::block.value(): {
-        size_t arity = 0;
-        size_t param_arity = 0;
-        auto& args = instruction.arguments().get<Instruction::StructuredInstructionArgs>();
-        switch (args.block_type.kind()) {
-        case BlockType::Empty:
-            break;
-        case BlockType::Type:
-            arity = 1;
-            break;
-        case BlockType::Index: {
-            auto& type = configuration.frame().module().types()[args.block_type.type_index().value()];
-            arity = type.results().size();
-            param_arity = type.parameters().size();
-        }
+    nops_to_remove.reverse();
+    auto nops_to_remove_copy = nops_to_remove;
+    size_t offset_accumulated = 0;
+    for (size_t i = 0; i < result.dispatches.size(); ++i) {
+        if (result.dispatches[i].instruction->opcode() == Instructions::nop) {
+            offset_accumulated++;
+            nops_to_remove.take_last();
+            continue;
         }
 
-        configuration.label_stack().append(Label(arity, args.end_ip, configuration.value_stack().size() - param_arity));
-        return;
-    }
-    case Instructions::loop.value(): {
-        auto& args = instruction.arguments().get<Instruction::StructuredInstructionArgs>();
-        size_t arity = 0;
-        if (args.block_type.kind() == BlockType::Index) {
-            auto& type = configuration.frame().module().types()[args.block_type.type_index().value()];
-            arity = type.parameters().size();
+        auto& args = result.dispatches[i].instruction->arguments();
+        if (auto ptr = args.get_pointer<Instruction::StructuredInstructionArgs>()) {
+            auto offset_to = [&](InstructionPointer ip) {
+                size_t offset = 0;
+                for (auto nop_ip : nops_to_remove.in_reverse()) {
+                    if (nop_ip < ip.value())
+                        ++offset;
+                    else
+                        break;
+                }
+                return offset;
+            };
+
+            InstructionPointer end_ip = ptr->end_ip.value() - offset_accumulated - offset_to(ptr->end_ip - ptr->else_ip.has_value());
+            auto else_ip = ptr->else_ip.map([&](InstructionPointer const& ip) -> InstructionPointer { return ip.value() - offset_accumulated - offset_to(ip - 1); });
+            auto instruction = *result.dispatches[i].instruction;
+            instruction.arguments() = Instruction::StructuredInstructionArgs {
+                .block_type = ptr->block_type,
+                .end_ip = end_ip,
+                .else_ip = else_ip,
+            };
+            result.extra_instruction_storage.append(move(instruction));
+            result.dispatches[i].instruction = &result.extra_instruction_storage.unsafe_last();
```

Comment:
is patching in place a No-No?
as this introduces quite the non linearity in the memory access pattern

Reply:
As in patching the input instruction stream? it would break wasm's `print` stuff by showing things that aren't necessarily there so I'm somewhat against it.

current bottlenecks are above this stuff anyhow, so I'm not too worried about it yet.